src/main/scala/com/fulcrumgenomics/bam/ConsensusCaller.scala

```diff
@@ -0,0 +1,369 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam
+
+import com.fulcrumgenomics.bam.ConsensusCallerOptions._
+import com.fulcrumgenomics.util.PhredValue.{OneProbability, ZeroPhred, ZeroProbability}
+import com.fulcrumgenomics.util.{PhredValue, ProgressLogger}
+import dagr.commons.CommonsDef._
+import htsjdk.samtools._
+import htsjdk.samtools.util.SequenceUtil
+
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+object ConsensusCallerOptions {
+  /** Various default values for the consensus caller. */
+  val DefaultAttribute: String                       = "RX"
+  val DefaultErrorRatePreUmi: PhredValue             = 50
+  val DefaultErrorRatePostUmi: PhredValue            = 0
+  val DefaultMaxBaseQuality: PhredValue              = 40
+  val DefaultBaseQualityShift: PhredValue            = 10
+  val DefaultMinConsensusBaseQuality: PhredValue     = 13
+  val DefaultMinReads: Int                           = 1
+  val DefaultMinMeanConsensusBaseQuality: PhredValue = 13
+}
+
+/** Holds the parameters/options for consensus calling. */
+case class ConsensusCallerOptions(attribute: String                       = DefaultAttribute,
+                                  errorRatePreUmi: PhredValue             = DefaultErrorRatePreUmi,
+                                  errorRatePostUmi: PhredValue            = DefaultErrorRatePostUmi,
+                                  maxBaseQuality: PhredValue              = DefaultMaxBaseQuality,
+                                  baseQualityShift: PhredValue            = DefaultBaseQualityShift,
+                                  minConsensusBaseQuality: PhredValue     = DefaultMinConsensusBaseQuality,
+                                  minReads: Int                           = DefaultMinReads,
+                                  minMeanConsensusBaseQuality: PhredValue = DefaultMinMeanConsensusBaseQuality
+                                 )
+
+object ConsensusCaller {
+  /** Creates a consensus read from the given read and quality string tuples.  If no consensus read was created, None is returned. */
+  def fromBasesAndQualities(basesAndQualities: Seq[(String, String)],
+                            options: ConsensusCallerOptions = new ConsensusCallerOptions()
+                           ): Option[(String, String)] = {
+    if (basesAndQualities.exists { case (b, q) => b.length != q.length }) {
+      throw new IllegalArgumentException("Found a read with the bases and qualities having different length")
+    }
+    // check to see if we have enough reads.
+    if (basesAndQualities.length < options.minReads) return None
+
+    // extract the bases and qualities, and adjust the qualities based on the given options.
+    val bases: Seq[String] = basesAndQualities.map { _._1.toUpperCase }
```

Comment:
While true, it confused me for a minute that `bases` is the bases from all the (likely multiple) reads, not just the bases from a single read.